src/evict/evict_lru.c

```diff
@@ -1057,17 +1058,31 @@ retry:	while (slot < max_entries && ret == 0) {
 			WT_ASSERT(session, session->split_gen == 0);
 		}
 
-		__wt_spin_unlock(session, &cache->evict_walk_lock);
-
 		/*
-		 * If we didn't find any candidates in the file, skip it next
-		 * time.
+		 * If we didn't find any candidates in the file, mark it to be
+		 * skipped for a few runs and remove it and insert to the tail.
 		 */
-		if (slot == prev_slot)
-			btree->evict_walk_period = WT_MIN(
-			    WT_MAX(1, 2 * btree->evict_walk_period), 100);
-		else
-			btree->evict_walk_period = 0;
+		if (slot == prev_slot) {
+skip:			if (dhandle->evict_empty_passes++ > 100 ) {
+				dhandle->evict_skip_until = laps + 15;
+				STAILQ_REMOVE(&conn->dhlh,
+				    dhandle, __wt_data_handle, l);
+				STAILQ_INSERT_TAIL(&conn->dhlh, dhandle, l);
+			}
+		} else {
+			dhandle->evict_empty_passes = 0;
+			/* If we find something of value, move it back up. */
+			if (dhandle->evict_skip_until > laps) {
```

Comment:
This seems weird to me: I don't think we should visit files if we've set their `skip_until` in the future.